File: np-lib/shared/sh_cacheable.js
function Cacheable(getValueCb, options) {
  const key = '_';
  const map = CacheableMap((ctx, _, ...args) => {
    return getValueCb(ctx, ...args);
  }, options);

  function getCachedValue(...args) {
    return map(key, ...args);
  }

  getCachedValue.reset = function() {
    map.reset(key);
  }

  return getCachedValue;
}

function CacheableMap(getValueCb, options) {
  const ttl = options.timeToLive || 60000;
  const cachedValues = {};

  async function ensureCachedValue(key, ...args) {
    let cachedValue = cachedValues[key];
    if (!cachedValue) {
      cachedValue = { value: null, lastUpdated: 0 };
      cachedValues[key] = cachedValue;
    }

    const now = Date.now();
    if (cachedValue.lastUpdated === 0 || now - cachedValue.lastUpdated > ttl) {
      const [shouldCache, value] = await getValueCb(cachedValue, key, ...args);
      if (shouldCache) {
        cachedValue.lastUpdated = now
        cachedValue.value = value
      }
      return value;
    }

    return cachedValue.value;
  }

  async function getCachedValue(key, ...args) {
    return await ensureCachedValue(key, ...args);
  }

  getCachedValue.reset = function(key) {
    const cachedValue = cachedValues[key];
    if (cachedValue) cachedValue.lastUpdated = 0;
  }

  return getCachedValue;
}

File: np-lib/shared/sh_cacheable.lua
function Cacheable(getValueCb, options)
  local key = 1
  local map = CacheableMap(function (ctx, _, ...)
    return getValueCb(ctx, ...)
  end, options)

  return {
    get = function (...) return map.get(key, ...) end,
    reset = function () map.reset(key) end
  }
end

function CacheableMap(getValueCb, options)
  local self = {}
  local ttl = options.timeToLive or 60000
  local cachedValues = {}

  local function ensureCachedValue(key, ...)
    local cachedValue = cachedValues[key]
    if cachedValue == nil then
      cachedValue = { value = nil, lastUpdated = 0 }
      cachedValues[key] = cachedValue
    end

    local now = GetGameTimer()
    if cachedValue.lastUpdated == 0 or now - cachedValue.lastUpdated > ttl then
      local shouldCache, value = getValueCb(cachedValue, key, ...)
      if shouldCache then
        cachedValue.lastUpdated = now
        cachedValue.value = value
      end
      return value
    end

    return cachedValue.value
  end

  function self.get(key, ...)
    return ensureCachedValue(key, ...)
  end

  function self.reset(key)
    local cachedValue = cachedValues[key]
    if cachedValue ~= nil then cachedValue.lastUpdated = 0 end
  end

  return self
end
